pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -550,6 +580,16 @@ impl InMemoryLayer {
             for (_blk, lsn, _pv) in inner.page_versions.ordered_page_version_iter(None) {
                 assert!(lsn <= end_lsn);
             }
+
+            // It's a bit premature to subtract the global mem usage here already.
+            // This layer consumes memory until it's written out to disk and dropped.
+            // But GLOBAL_OPEN_MEM_USAGE is used to trigger layer eviction, if there are
+            // too many open layers, and from that point of view this should no longer be
+            // counted against the global mem usage.
```

Comment:
Why should it not count against the global memory usage until `Drop`d? Other readers could hold on to the Arc for a while, and we already have the `Drop` impl

Reply:
Yeah, it's definitely stilll consuming memory, even when it's frozen. However, if we don't reduce GLOBAL_OPEN_MEM_USAGE here, then another WAL receiver thread would think that it needs to also evict a layer to bring down memory usage, even though a layer is already being evicted and the memory usage will be below the threshold as soon as the eviction has finished.

For example, imagine that there are 10 timelines active in the system, receiving WAL. GLOBAL_OPEN_MEM_USAGE reaches the MAX_OPEN_MEM threshold, and one of the WAL receivers starts eviction. While it's busy writing the victim layer to disk, another WAL receiver also sees that we are above the threshold, and *also* starts the eviction process, evicting another layer. Third WAL receiver comes along and does the same, and so forth. In the end, all ten receivers stopped to evict a layer, even though just evicting one layer would have been enough to bring down the memory usage. By subtracting MAX_OPEN_MEM earlier, before the expensive part where the layer is written to disk, we reduce the window for that. The race condition is still there, but it's much smaller. (This doesn't need to be exact, it's relatively harmless if you evict some layers a little bit earlier than necessary, but we should try to avoid a "thundering herd" problem)

Perhaps we should have two counters, one for the memory used by open layers, and another for in-memory layers that have been frozen but not evicted to disk yet. The former would be used to decide when to start the eviction process, the latter would be nice to know for monitoring. But I think that can be a separate PR later.